### DISCUSSION TITLE
Our recent works on making copr sign process more effective

### DISCUSSION THREAD

**Discussion by pkking** [fedora-copr/copr#2935] [Show and tell]
Hi team, im very excited to share our recent works on copr and looking forward to hear you opinions :)

# Why we need a faster sign process?
Once upon a while we have to sign a lot of packages in our copr instances(called EUR) because we need to switch to a new signing key.
When i courageously run the `copr_fix_gpg.py` for my first 2000 packages, the whole process halt for about 45mins and meanwhile i am so worried it will jam forever when i then start to sign the left 20,000 packages.

After this heartquake signing process, i take some time to investigate the copr sign process.

## How copr sign a rpm
![image](https://github.com/fedora-copr/copr/assets/2853724/c60b039f-fc83-461c-ac6d-0d4ccbf8304e)

Here's a image illustrating how copr sign a rpm:

On the client-side(copr-backend):
1. backend called a RESTful API `/gen_key` to gen the key when new project is created/forked.
2. backend called the `sign` tool in `obs-signd` rpm package to sign each rpm separately.

On the server-side(copr-keygen):
1. once the `/gen_key` API is called, copr-keygen will call `pgp2` to generate a keypair and store it in the database.
2. when backend calls `sign` tool to sign a rpm, the tool will establish a TCP connection with `obs-signd` running on keygen node and send the rpm `payload` to it.

## Why backend sign not so fast
1. backend call `sign` to sign rpm one by one, actually each sign process were irrelevant
2. `sign` can't resign a rpm, so we must `unsign && sign` the rpm each time
3. `obs-signd` also [do not support concurrent](https://github.com/openSUSE/obs-sign/issues/45)

# Way to speed up
At the early of 2023, openEuler community open sourced [signatrust](https://gitee.com/openeuler/signatrust) as a highly secure, async and efficient solution for signing Linux packages and binaries platform.

Signatrust itself provides secure and high throughput for all signing requests, and that's what we thought a mordern signing infrastructure should be. Here

## Why signatrust is faster
1. the sign client start multi workers to sign rpms parallelly
2. signatrust provided a more quickly API to check key existence
3. signatrust will construct the rpm with signature in place, so we do not need unsigned the rpm before signing
4. each sign request will be handled in async way, so multiple requests can be processed concurrently

## Benchmarks
Here are some benchmarks comparing the performance of signatrust vs the obs-signd approach:

![image](https://github.com/fedora-copr/copr/assets/2853724/3c169f9d-febd-4596-b126-c3afe3d411e5)

Lets explaine the benchmarks:
1. both obs-signd and signatrust runs on a 8u8G kubernetes pod and within same namespace with `copr-backend`
2. we tried three kind of workload to simulate most our daily works on copr:
  - sign few package: most of the build runs workload like this
  - sign some packages: most of the project forked or batch rebuild
  - sign massive packages: some rare but important SOP/maintaince case

The bench results showed signatrust were 3x-10x faster than obs-signd in all three cases. This is mainly because:
1. signatrust client used multi-threading to sign rpms in parallel, thats why in few package case it was 3x faster
1. signatrust will not unsign rpm before signing, this avoids unnecessary rpm repackaging
1. signatrust server using async io model to handle multiple signing requests concurrently

And also, we monitord obs-signd CPU and Memory usage, it's always very low during signing process. This indicates obs-signd is not fully utilizing the server resources and not handling requests concurrently.

![image](https://github.com/fedora-copr/copr/assets/2853724/f566d4cb-6a57-4f07-ade3-03cb96c2204d)

# What now
openEuler copr instance EUR(openEuler User Repo) have already finished migrating our sign backend service from `obs-signd` to `signatrust`

To sum up in a word: each time the user build a single package on a chroot cost less 5~10 seconds than before

see the i3 builds [using obs-signd](https://eur.openeuler.openatom.cn/results/mywaaagh_admin/i3wm/openeuler-23.03-x86_64/00090569-i3/backend.log.gz) vs [using signatrust](https://eur.openeuler.openatom.cn/results/@Infrastructure/i3wm/openeuler-23.03-x86_64/00091763-i3/backend.log.gz)

Still we patched backend to make sign backend plugable, if you were interested, i'd like make a PR :)

**Comment by praiskup** [comment]
Thank you for the analysis, @pkking!

One missing point we should consider is signing of large (S)RPMs.  Consider that the RPMs are often larger than a few MBs, and sometimes are also very large >= 1GB (we have an internal use-case where RPMs are >= 3GB).  The `obs-sign` software is very efficient here WRT the client-server approach (the client calculates the hashing part of the signature, so the whole large RPM blob doesn't have to be sent to the server for actual signing).  Have you tried to measure this use-case and compare?

> Still we patched backend to make sign backend plugable, if you were interested, i'd like make a PR :)

Definitely.  As long as it is opt-in, it's definitely OK to have the support upstream!

**Reply by pkking** [reply]
Oh, this case is not covered by our test cases, will try to test and replay later :)

**Reply by TommyLike** [reply]
@praiskup currently we only calculate the rpm digest header at the client side, please check [this](https://gitee.com/openeuler/signatrust/blob/master/src/client/file_handler/rpm.rs#L121) , the payload and header will be transferred to the server side cause it will be used for the PGP library directly, not sure about the obs-sign detail although, considering the signing system and COPR will be deployed in the same network area, I am wondering what is the possibility of having rpm larger than 1GB in the COPR environment and what exactly the impact of these, we will try to find this, thanks.

**Reply by Conan-Kudo** [reply]
There are several such examples in Fedora. `chromium`, `texlive`, and `0ad-data` are good test cases.

**Reply by TommyLike** [reply]
@Conan-Kudo Thanks we will test these.

**Reply by TommyLike** [reply]
@praiskup there is another point, considering the difference of rpm signature v3 and v4 there would be big performane improvement if no payload is transfered on the network, so, what's the strategy for COPR to stop support signature v3? check this: https://github.com/rpm-software-management/rpm/issues/863

**Comment by praiskup** [comment]
@pkking WRT the mass-resign process, note that we plan to [use PULP](https://github.com/fedora-copr/copr/issues/2533) as the RPM repository manager in the future.  So you might want to work on integrating Signatrust with PULP, too?

**Reply by TommyLike** [reply]
@praiskup  Just checked the document of pulp and pulp rpm, I think it does not support signing the rpm file but only the repo metadata file currently?

**Reply by ipanova** [reply]
That's correct, only metadata files for now.

**Comment by pkking** [comment]
> @pkking WRT the mass-resign process, note that we plan to [use PULP](https://github.com/fedora-copr/copr/issues/2533) as the RPM repository manager in the future. So you might want to work on integrating Signatrust with PULP, too?

Its my fisrt glance on PULP, it seems to be a funny project who also happened to solv many of our problems and i will talk to signatrust team to discuss the integration :)

**Reply by Conan-Kudo** [reply]
cc: @bmbouter and @dralley for this discussion. They can provide useful information about Pulp. :)

**Reply by ipanova** [reply]
@pkking at the moment we have captured in a ticket [native obs-signd support for signing Pulp software](https://github.com/pulp/pulp_rpm/issues/2986), we are opened to look into Signatrust if @praiskup  and rest think COPR will benefit from it. A pulp ticket with detailed info is more than welcome and as always if you feel like you have the desire and motivation to contribute we will be more than happy to assist you through questions.

**Reply by praiskup** [reply]
I haven't done an analysis myself, so I trust @pkking that it performs nicely and is worth it (even if only their instance used it).  I just miss the analysis of signing rather big RPMs (see above).

Also, once Copr is on PULP and RPMs are stored somewhere in `s3`, re-signing effectively (what @pkking's team seems to do heavily) will be an interesting problem to solve even with obs-sign.

**Reply by dralley** [reply]
Yes, realistically I don't see that being especially practical with header-and-payload signatures.  But even with header-only signatures, we'd still need to persist a checksum of the header which we don't currently do, OR download the header bytes using a range request for processing.

**Comment by TommyLike** [comment]
@praiskup @Conan-Kudo 

We conducted performance tests in two scenarios:

**Scenario 1**: Common scenarios for RPM package signing. The test files range from a few MB to several GB. There are 4,000 RPM packages occupying 19 GB of disk space in total. All RPM files can be downloaded at: https://repo.openeuler.org/openEuler-22.03-LTS/source/Packages/

**Scenario 2**: Extreme scenario where all RPM packages are large files. Actually each RPM has a size of 1.8 GB, with a total of 50 files, occupying 100 GB of disk space. The test RPM package can be downloaded at: https://dl.fedoraproject.org/pub/fedora/linux/releases/38/Everything/source/tree/Packages/s/speed-dreams-2.2.3-6.fc38.src.rpm

**Test Environment**:
Each test consists of one client and one server, with each instance restricted to 8 CPU cores and 8 GB of RAM. The client runs with a concurrency of 8. Note that the obs-sign client does not support simultaneous signing. To address this, we used a script to simultaneously launch 8 obs-sign clients for signing.

**Test Results**:
**obs-sign(v3 signature):**
**Scenario 1**: 20 minutes and 51 seconds
**Scenario 2**: 18 minutes and 5 seconds

**signatrust(v3 signature):**
**Scenario 1**: 4 minute and 43 seconds
**Scenario 2**: OOM

**signatrust(v4 signature):**
**Scenario 1**: 1 minute and 53 seconds
**Scenario 2**: 18 minutes and 34 seconds

**CPU and Memory Consumption for Scenario 2**
**obs-sign:**
Client CPU:
![image](https://github.com/fedora-copr/copr/assets/5671225/b51594c2-e718-45e2-818d-2bfd42ac1b78)

Client Memory:
![image](https://github.com/fedora-copr/copr/assets/5671225/66930100-8b1d-4cf2-9d72-448e8afac3ff)

Server CPU:
![image](https://github.com/fedora-copr/copr/assets/5671225/93b18ad8-1ac1-4afc-a75c-bbb409c99558)

Server Memory:
![image](https://github.com/fedora-copr/copr/assets/5671225/378dc569-1973-4a44-9865-ee6dfd442da3)

**signatrust:**
Client CPU:
![image](https://github.com/fedora-copr/copr/assets/5671225/be595786-55bd-4764-8c0c-4fc0276ada8b)

Client Memory:
![image](https://github.com/fedora-copr/copr/assets/5671225/8bda3377-6cff-4cae-8b9d-e16e6e49a7d9)

Server CPU:
![image](https://github.com/fedora-copr/copr/assets/5671225/06658eb5-ebeb-49b2-af49-3b09a28c0663)

Server Memory:
![image](https://github.com/fedora-copr/copr/assets/5671225/0f07fccf-11b1-429c-8afd-421f8dc54371)

**Conclusion**:
In common scenarios, Signatrust significantly improves the efficiency of RPM package signing by taking advantage of asynchronous processing. 
In the extreme scenario of large RPM files. obs-sign, due to its approach of only transmitting digests for signature computation, it can reduce network consumption and improve the efficiency for RPM v3 signature, but when compared to the RPM V4 Signature mode that only signs the header, the overall efficiency is similar.

**Reply by praiskup** [reply]
Thank you for taking a look. I miss the **obs-sign(v4 signature)** part, don't you have results also from there? I suppose that Signatrust is much faster with v4 signatures?  We have the https://github.com/fedora-copr/copr/issues/2757 btw, which seems to be caused by `-4` option, right?

**Reply by TommyLike** [reply]
if obs-sign will send digest for v3 signature, I guess the v4 signature will be a little(not too much) faster because the calclucation is not needed, but we will have a try. @praiskup

**Reply by dralley** [reply]
It's maybe best not to talk about it in terms of v3 signatures and v4 signatures since that overlaps with PGP v3 and v4, which is actually entirely unrelated.

There are header-only signatures, and header-and-payload signatures, where the latter is also sometimes known as RPMv3 signatures.

On recent distros, RPM does not produce header-and-payload / RPMv3 signatures by default https://github.com/rpm-software-management/rpm/pull/1093

Thus header-and-payload signatures ought to cease being particularly relevant in a few months when EL7 goes end-of-life.  I assume at that point COPR would start dropping EL7 support?

**Reply by TommyLike** [reply]
thanks for your explaination. @dralley

**Comment by dralley** [comment]
Which I suppose you mentioned here already https://github.com/fedora-copr/copr/discussions/2935#discussioncomment-7221058
